File: game/src/core/ghostcodec.ts
// Ghost channel codec: delta + zigzag + LEB128 varint + base64url (GAME_DESIGN.md §8.2). Owner: O3.
// Must match game/tools/ghostcodec.py (tests/fixtures/ghostcodec_vectors.json).
//
//   quantise : x -> round(x * 1e4) (0.1 mm), th -> round(th * 1e4), F -> round(F * 10)   (Math.round)
//   layout   : first value as is, then differences to the previous value
//   varint   : zigzag (n >= 0 ? 2n : -2n-1), unsigned LEB128 (7 bits, low group first)
//   text     : base64url without padding
//
// Arithmetic uses plain numbers (not 32-bit bit operations) so values up to 2^52 survive.

export type GhostChannel = 'x' | 'th' | 'f';
export const CHANNEL_SCALE: Readonly<Record<GhostChannel, number>> = { x: 1e4, th: 1e4, f: 10 };

const B64 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_';
const B64_INV: Int16Array = (() => {
  const t = new Int16Array(128).fill(-1);
  for (let i = 0; i < 64; i++) t[B64.charCodeAt(i)] = i;
  return t;
})();

function b64urlFromBytes(bytes: ArrayLike<number>, len: number): string {
  let out = '';
  let i = 0;
  for (; i + 2 < len; i += 3) {
    const n = (bytes[i]! << 16) | (bytes[i + 1]! << 8) | bytes[i + 2]!;
    out += B64[(n >> 18) & 63]! + B64[(n >> 12) & 63]! + B64[(n >> 6) & 63]! + B64[n & 63]!;
  }
  const rest = len - i;
  if (rest === 1) {
    const n = bytes[i]! << 16;
    out += B64[(n >> 18) & 63]! + B64[(n >> 12) & 63]!;
  } else if (rest === 2) {
    const n = (bytes[i]! << 16) | (bytes[i + 1]! << 8);
    out += B64[(n >> 18) & 63]! + B64[(n >> 12) & 63]! + B64[(n >> 6) & 63]!;
  }
  return out;
}

function bytesFromB64url(s: string): Uint8Array {
  if (s.length % 4 === 1) throw new Error('ghostcodec: bad base64url length');
  const out = new Uint8Array(Math.floor((s.length * 3) / 4));
  let o = 0;
  let acc = 0;
  let bits = 0;
  for (let i = 0; i < s.length; i++) {
    const c = s.charCodeAt(i);
    const v = c < 128 ? B64_INV[c]! : -1;
    if (v < 0) throw new Error(`ghostcodec: bad base64url character at ${i}`);
    acc = ((acc << 6) | v) & 0xffffff;
    bits += 6;
    if (bits >= 8) {
      bits -= 8;
      out[o++] = (acc >> bits) & 0xff;
    }
  }
  return out.subarray(0, o);
}

/** Quantises a channel with Math.round (ties towards +inf, like tools/ghostcodec.py js_round). */
export function quantize(values: ArrayLike<number>, channel: GhostChannel): number[] {
  const k = CHANNEL_SCALE[channel];
  const out = new Array<number>(values.length);
  for (let i = 0; i < values.length; i++) out[i] = Math.round(values[i]! * k) + 0; // + 0 turns -0 into 0
  return out;
}

/** Quantised integers -> base64url string. */
export function encodeChannel(values: ArrayLike<number>): string {
  const bytes: number[] = [];
  let prev = 0;
  for (let i = 0; i < values.length; i++) {
    const n = Math.round(values[i]!);
    const d = i === 0 ? n : n - prev;
    prev = n;
    let z = d >= 0 ? 2 * d : -2 * d - 1;
    for (;;) {
      const low = z % 128;
      z = Math.floor(z / 128);
      if (z > 0) {
        bytes.push(low | 0x80);
      } else {
        bytes.push(low);
        break;
      }
    }
  }
  return b64urlFromBytes(bytes, bytes.length);
}

/** base64url string -> quantised integers. Throws on malformed input (bad characters, truncated varint). */
export function decodeChannel(s: string): Int32Array {
  const data = bytesFromB64url(s);
  const out: number[] = [];
  let prev = 0;
  let z = 0;
  let mul = 1;
  let pending = false;
  for (let i = 0; i < data.length; i++) {
    const byte = data[i]!;
    z += (byte & 0x7f) * mul;
    if (byte & 0x80) {
      mul *= 128;
      pending = true;
      if (mul > 2 ** 49) throw new Error('ghostcodec: varint too long');
      continue;
    }
    const d = z % 2 === 0 ? z / 2 : -(z + 1) / 2;
    prev = out.length === 0 ? d : prev + d;
    out.push(prev);
    z = 0;
    mul = 1;
    pending = false;
  }
  if (pending) throw new Error('ghostcodec: truncated varint');
  return Int32Array.from(out);
}

/** decodeChannel followed by division by the channel scale. */
export function decodeChannelValues(s: string, channel: GhostChannel): Float64Array {
  const ints = decodeChannel(s);
  const k = CHANNEL_SCALE[channel];
  const out = new Float64Array(ints.length);
  for (let i = 0; i < ints.length; i++) out[i] = ints[i]! / k;
  return out;
}

/** quantize + encodeChannel. */
export function encodeChannelValues(values: ArrayLike<number>, channel: GhostChannel): string {
  return encodeChannel(quantize(values, channel));
}

File: game/tools/ghostcodec.py
"""Ghost channel codec (spec §8.2 "チャンネルの符号化"); mirrored by src/core/ghostcodec.ts.

  quantise : x -> round(x * 1e4)   (0.1 mm)
             th -> round(th * 1e4) (0.1 mrad)
             f -> round(F * 10)    (0.1 N)
             round() is JS Math.round (half up, towards +inf) so both sides agree.
  layout   : first value as is, then differences to the previous value
  varint   : zigzag (n >= 0 ? 2n : -2n-1), then unsigned LEB128 (7 bits, low first)
  text     : base64url without padding
"""

from __future__ import annotations

import base64
import math

SCALE = {"x": 1e4, "th": 1e4, "f": 10.0}


def js_round(v: float) -> int:
    """Math.round: floor(v + 0.5) (ties go towards +inf, -0.5 -> -0)."""
    return int(math.floor(v + 0.5))


def zigzag(n: int) -> int:
    return 2 * n if n >= 0 else -2 * n - 1


def unzigzag(z: int) -> int:
    return z >> 1 if (z & 1) == 0 else -((z + 1) >> 1)


def b64url_encode(b: bytes) -> str:
    return base64.urlsafe_b64encode(b).rstrip(b"=").decode("ascii")


def b64url_decode(s: str) -> bytes:
    return base64.urlsafe_b64decode(s + "=" * (-len(s) % 4))


def encode_ints(ints) -> str:
    out = bytearray()
    prev = 0
    for i, n in enumerate(ints):
        n = int(n)
        d = n if i == 0 else n - prev
        prev = n
        z = zigzag(d)
        while True:
            byte = z & 0x7F
            z >>= 7
            if z:
                out.append(byte | 0x80)
            else:
                out.append(byte)
                break
    return b64url_encode(bytes(out))


def decode_ints(s: str) -> list[int]:
    data = b64url_decode(s)
    out, prev, z, shift = [], 0, 0, 0
    for byte in data:
        z |= (byte & 0x7F) << shift
        if byte & 0x80:
            shift += 7
            continue
        d = unzigzag(z)
        prev = d if not out else prev + d
        out.append(prev)
        z, shift = 0, 0
    if shift:
        raise ValueError("truncated varint")
    return out


def quantize(values, channel: str) -> list[int]:
    k = SCALE[channel]
    return [js_round(float(v) * k) for v in values]


def dequantize(ints, channel: str) -> list[float]:
    k = SCALE[channel]
    return [n / k for n in ints]


def encode_channel(values, channel: str) -> str:
    return encode_ints(quantize(values, channel))


def decode_channel(s: str, channel: str) -> list[float]:
    return dequantize(decode_ints(s), channel)


if __name__ == "__main__":
    import random
    rng = random.Random(0)
    for _ in range(2000):
        n = rng.randint(0, 50)
        ints = [rng.randint(-2**27, 2**27) for _ in range(n)]
        assert decode_ints(encode_ints(ints)) == ints
    assert js_round(-0.5) == 0 and js_round(0.5) == 1 and js_round(-1.5) == -1
    print("ghostcodec self-test ok")
